### ISSUE TITLE
"smooth_damp" functions for primitives and math structs

### ISSUE THREAD

**Issue by vveisard** [bevyengine/bevy#13408]
## What problem does this solve or what need does it fill?

I need to smoothly move a value towards a target value over time using velocity and other parameters (see below).

## What solution would you like?

I would like a set of "smooth_damp" functions for primitives and which have:
- current
- target
- velocity
- smooth time
- max value delta
- time delta

Importantly, I would also like a *spherical* smooth_damp for Vec2 and Vec3.

## What alternative(s) have you considered?

A similar effect can be achieved using interpolation functions while updating the start value. However, this lacks all the parameters of a smooth_damp function.

## Additional context

https://docs.unity3d.com/ScriptReference/Vector3.SmoothDamp.html

**Comment by alice-i-cecile** [comment]
I definitely want this: very useful!

This feels like something that could be added as an auto-implemented method on the `Animatable` trait. Does that make sense to you?

You would probably need to special case the spherical interpolation in some form. Perhaps a newtyped Vec2/Vec3 that captures the notion of operating in spherical coordinates?

**Comment by IQuick143** [comment]
`Perhaps a newtyped Vec2/Vec3 that captures the notion of operating in spherical coordinates?`
How about a Dir2/3, it's already a newtyped Vec2/3, bound to the unit sphere.

**Comment by SIGSTACKFAULT** [comment]
If `Dir3` gets it then `Quat` should too

**Comment by vveisard** [comment]
> This feels like something that could be added as an auto-implemented method on the `Animatable` trait. Does that make sense to you?

Yes, `Animatable` seems like the right place for a `smooth_damp` method. However, I would expect a `smooth_damp` method for `f32`, `Rotation2d`, `Vec2`, `Vec3`, `Dir2`, `Dir3`, and `Quat`, and some of those lack an `Animatable` implementation.

I'm not familiar enough with Rust to know if auto-implement is the right thing to do.

**Comment by alice-i-cecile** [comment]
Pretty much all of those should have an `Animatable` implementation :) But yes, this can just be a method on the trait with a default implementation of the method.
